src/qibo/transpiler/star_connectivity.py

```diff
@@ -0,0 +1,127 @@
+                as the original but respects the hardware connectivity.
+            hardware_qubits (list): List that maps logical to hardware qubits.
+                This is required for transforming final measurements.
+        """
+
+        middle_qubit = self.middle_qubit
+        # find the number of qubits for hardware circuit
+        if circuit.nqubits == 1:
+            nqubits = 1
+        else:
+            nqubits = max(circuit.nqubits, middle_qubit + 1)
+        # new circuit object that will be compatible with hardware connectivity
+        new = circuit.__class__(nqubits)
+        # list to maps logical to hardware qubits
+        hardware_qubits = list(range(nqubits))
+
+        # find initial qubit mapping
+        for i, gate in enumerate(circuit.queue):
+            if len(gate.qubits) == 2:
+                if middle_qubit not in gate.qubits:
+                    new_middle = find_connected_qubit(
+                        gate.qubits, circuit.queue[i + 1 :], hardware_qubits
+                    )
+                    hardware_qubits[middle_qubit], hardware_qubits[new_middle] = (
+                        hardware_qubits[new_middle],
+                        hardware_qubits[middle_qubit],
+                    )
+                break
+
+        # the first SWAP is not needed as it can be applied via virtual mapping
+        add_swap = False
+        for i, gate in enumerate(circuit.queue):
+            # map gate qubits to hardware
+            qubits = tuple(hardware_qubits.index(q) for q in gate.qubits)
+            if isinstance(gate, gates.M):
+                new_gate = gates.M(*qubits, **gate.init_kwargs)
+                new_gate.result = gate.result
+                new.add(new_gate)
+                continue
+
+            if len(qubits) > 2:
+                raise ConnectivityError(
+                    "Gates targeting more than two qubits are not supported."
+                )
+
+            elif len(qubits) == 2 and middle_qubit not in qubits:
+                # find which qubit should be moved to 0
+                new_middle = find_connected_qubit(
+                    qubits, circuit.queue[i + 1 :], hardware_qubits
+                )
+                # update hardware qubits according to the swap
+                hardware_qubits[middle_qubit], hardware_qubits[new_middle] = (
+                    hardware_qubits[new_middle],
+                    hardware_qubits[middle_qubit],
+                )
+                if add_swap:
+                    new.add(gates.SWAP(middle_qubit, new_middle))
+                # update gate qubits according to the new swap
+                qubits = tuple(hardware_qubits.index(q) for q in gate.qubits)
+
+            # add gate to the hardware circuit
+            if isinstance(gate, gates.Unitary):
+                # gates.Unitary requires matrix as first argument
+                backend = NumpyBackend()
+                matrix = gate.matrix(backend)
```

Comment:
This makes this part of the code backend-independent and also fix the tests for GPU backends

```suggestion
                matrix = gate.init_args[0]
```